crates/btc-relay/src/lib.rs

```diff
@@ -1218,6 +1316,70 @@ impl<T: Config> Pallet<T> {
         Ok(())
     }
 
+    fn is_reorg_required(
+        prev_blockchain_id: u32,
+        current_position: u32,
+        prev_height: u32,
+        current_height: u32,
+    ) -> Result<bool, DispatchError> {
+        if Self::is_reorg_according_to_chainwork() {
+            let previous_block_hash = Self::get_block_hash(prev_blockchain_id, prev_height)?;
+            let current_chain_id = Self::get_chain_id_from_position(current_position)?;
+            let current_block_hash = Self::get_block_hash(current_chain_id, current_height)?;
+
+            let previous_block_chain_work = Self::get_block_chainwork(previous_block_hash).unwrap();
+            let current_block_chain_work = Self::get_block_chainwork(current_block_hash).unwrap();
+
+            if previous_block_chain_work >= current_block_chain_work {
+                Ok(true)
+            } else {
+                Ok(false)
+            }
+        }
+        // if chain work is not set, reorg according to longest chain
+        else {
+            Ok(prev_height.saturating_add(Self::get_stable_transaction_confirmations()) <= current_height)
+        }
+    }
+
+    pub fn calculate_chainwork(prev_block_hash: H256Le) -> Result<U512, DispatchError> {
+        // Retrieve the previous block's header
+        let prev_block_header = Self::get_block_header_from_hash(prev_block_hash)?;
+
+        // Extract the target from the previous block's header
+        let prev_block_target = U512::from(prev_block_header.block_header.target);
+
+        // Calculate 2^256
+        let max_target = U512::from(2).pow(U512::from(256));
+
+        // Calculate the chainwork
+        let chainwork = max_target.saturating_sub(prev_block_target).saturating_sub(ONE);
+
+        // Check if division is possible, and perform the division if so
+        if let Some(chainwork_done) = chainwork.checked_div(prev_block_target.saturating_add(ONE)) {
+            Ok(chainwork_done.saturating_add(ONE))
+        } else {
+            // Handle division error
+            Err(Error::<T>::DivisionError.into())
+        }
+    }
+
+    fn _set_chainwork_for_block(block_header: &BlockHeader) -> Result<(), DispatchError> {
+        //Check if chain work exist for prev block
+        ensure!(
+            Self::contains_chainwork(block_header.hash_prev_block),
+            Error::<T>::ChainWorkDoesNotExist
+        );
+
+        //calculate chain work for current block
+        let chain_work = Self::calculate_chainwork(block_header.hash)?;
+
+        //set chain work for the block
+        Self::set_block_chainwork(block_header.hash, chain_work);
```

Comment:
nit: I personally don't like getter/setter functions around storage variables if it literally on a wrapper around the `get`/`insert` since when reading the code, you have to check the function to see if any logic is hidden there.

Reply:
noted